torch/_inductor/cudagraph_trees.py

```diff
@@ -0,0 +1,863 @@
+                return False
+
+        if not self.check_liveness(
+            self.expected_dead_indices_before_graph, self.path_weakrefs
+        ):
+            return False
+
+        # the cudagraph managed tensors which died upon recording must also die upon
+        # this invocation. it is too late to check after we've replayed the graph,
+        # because we would have already written over their memory.
+        for idx in self.cudagraph_managed_idxs:
+            inputs[idx] = None
+
+        check(
+            self.check_liveness(
+                self.expected_dead_indices_after_graph, self.path_weakrefs
+            ),
+            lambda: "TODO: graph recording observed an input tensor deallocate during graph "
+            " recording that did not occur during replay. Please file an issue.",
+        )
+        return True
+
+    def num_descendants(self) -> int:
+        "Total number of descendents of this node"
+        num_desc = 0
+        for children in self.children.values():
+            for child in children:
+                num_desc += 1
+                num_desc += child.num_descendants()
+        return num_desc
+
+
+def get_cudagraph_segments(pool_id):
+    segments = torch.cuda.memory_snapshot()
+    return [segment for segment in segments if segment["segment_pool_id"] == pool_id]
+
+
+def check_memory_pool(pool_id, live_tensors):
+    unique_storages = {t.untyped_storage().data_ptr() for t in live_tensors}
+    segments = get_cudagraph_segments(pool_id)
+
+    for segment in segments:
+        addr = segment["address"]
+        for block in segment["blocks"]:
+            if block["state"] == "active_allocated":
+                check(
+                    addr in unique_storages,
+                    lambda: f"{addr} allocated but not in live storages",
+                )
+                unique_storages.remove(addr)
+
+            addr += block["size"]
+
+    check(
+        len(unique_storages) == 0,
+        f"These storage data ptrs are not allocated but should be {unique_storages}",
+    )
+
+
+class CUDAGraphTreeManager(object):
+    """
+    Groups individual recordings or executions of cuda graphs into a tree of recordings,
+    and checks required invariants.
+
+    When graphs are recorded in the same tree, it enforces subsequent execution
+    to follow the same order and have the same output tensor livespans. To remove
+    unnecessary coupling of cuda graphs (and additional imposed invariants),
+    the tree manager will end a currently recording tree whenever it is valid - when
+    the memory pool no longer has any live allocations.
+
+    We ignore outputs from a previous generation that correspond to prior model outputs.
+    Currently this is hardcoded `GenerationTracker.generation` tracked in torch dynamo.
+    # TODO: make generation increment configurable, warn on overwrite
+    """
+
+    def __init__(self):
+        self.roots: Dict[FunctionID, List[CUDAGraphNode]] = defaultdict(list)
```

Comment:
Is there a bound to how many CUDAGraphNodes in this list could there be?

Reply:
No bound, although you would expect the invariants to not change much

Reply:
I guess my question is, can we get into some sort of horrible situation where we just keep allocating new nodes without deallocating them.

Reply:
If you are in a sufficiently dynamic model there's not really a good way to avoid that. Could warn after some threshold. But as of now we're not always enabling it.